test/thread_test.c

```diff
@@ -0,0 +1,91 @@
+#include <threads.h>
+#include <stdio.h>
+#include <assert.h>
+#include <stdatomic.h>
+#include <stdbool.h>
+#include <time.h>
+
+#include "libisyntax.h"
+
+typedef struct test_thread_t {
+  thrd_t thread_id;
+  atomic_int* atomic_sync_flag; // Spinlock to sync threads.
+  bool force_sync;
+
+  // TODO(avirodov): implement if needed. Or we can keep making sure that test output comes after test work.
+  // char* output_buffer; // Avoid console output due to it may be implemented with a mutex or other sync.
+  // int output_buffer_len;
+  // int output_buffer_capacity;
+
+  void* arg;
+  int (*func)(void*);
+  int result;
+} test_thread_t;
+
+int test_print(void* arg) {
+  clock_t current_clock = clock();
+  printf("test_print tid=%ld currect_clock=%ld\n", thrd_current(), current_clock);
+  return 0;
+}
+
+int parallel_sync_and_call(void* arg) {
+  test_thread_t* test_thread = (test_thread_t*)arg;
+  // Wait in spinlock.
+  while (test_thread->force_sync && atomic_load(&test_thread->atomic_sync_flag) == 0) {
+    // noop.
+  }
+
+  // Run test code.
+  return test_thread->func(test_thread->arg);
+}
+
+void parallel_run(int (*func)(void*), void* arg, bool force_sync) {
+#define n_threads 10
+  const int n_iterations = 2;
+
+  for (int iter = 0; iter < n_iterations; ++iter) {
+    printf("== parallel run iter %d ==\n", iter);
+    atomic_int atomic_sync_flag = 0;
+    test_thread_t threads[n_threads] = {0};
+
+    // Spawn the threads.
+    for (int thread_i = 0; thread_i < n_threads; ++thread_i) {
+
+      threads[thread_i].atomic_sync_flag = &atomic_sync_flag;
+      threads[thread_i].func = func;
+      threads[thread_i].arg = arg;
+      threads[thread_i].force_sync = force_sync;
+
+      assert(thrd_create(&threads[thread_i].thread_id, parallel_sync_and_call, &threads[thread_i]) == thrd_success);
```

Comment:
Will get optimized out in Release builds.

Reply:
Fixed, thank you for catching this.